### ISSUE TITLE
Progress Messages for MPI

### ISSUE THREAD

**Issue by jacobseiler** [sage-home/sage-model#14]
In serial mode, a progress bar is displays the incremental progress (as a fraction of the number of trees processed).  However, running in MPI, there is no output messages at all.  As a result you can't be sure if the code is running or just hanging or if its slow or running fast...

The key line is [here](https://github.com/sage-home/sage-model/blob/2efa0ed78ebe0ebbfddcdbdd7bdcef3788e99c4d/src/sage.c#L114) where `progressbar` is only used if `NTasks == 1`.  A first solution to this issue is to remove this condition.  However, since we load balance on number of trees and not the number of halos, we could encounter a scenario where Task 0 has comparatively faster/slower computational time compared to other tasks.  This would lead to a progress bar that does not properly represent the computation across ALL tasks.

I'm not 100% sure on the best implementation as a result of this.

(Opening this as an issue rather than a ToDo item because I feel it is very important).

**Comment by manodeep** [comment]
The only way to correctly display a progressbar requires an MPI dependency (i.e., an `MPI_Barrier` followed by an `MPI_Sum`) within `libsage.a` we should avoid that unless absolutely required.

**Comment by manodeep** [comment]
We can always enable this via an `Makefile` option, say `-DSHOW_MPI_PROGRESSBAR`. That way, we can disable when necessary.

**Comment by jacobseiler** [comment]
Yes that's a good point.  I think for now I'll just use the hack solution (setting the `NTasks == 1` if condition) and then we can decide later if something more elegant is required.

**Comment by jacobseiler** [comment]
I have a vague recollection that you may have addressed this in one of your latest `illustris` commits @manodeep ?  

Regardless, should be addressed/resolved for `v1.0.0`.

**Comment by manodeep** [comment]
I have not implemented the `MPI_Barrier + MPI_Sum` needed to show the progress accurately when in MPI mode with multiple MPI tasks. It is quite trivial to add. 

I am still of the opinion that this accurate MPI progress bar should be disabled by default and only enabled when explicitly requested.

**Comment by manodeep** [comment]
I have however changed the logic such that the progress bar always goes up to the total number forests processed in both serial and MPI modes. Code [here](https://github.com/jacobseiler/sage-model/blob/illustris/src/sage.c#L113) and [here](https://github.com/jacobseiler/sage-model/blob/illustris/src/sage.c#L136)

**Comment by jacobseiler** [comment]
Yeah I think doing a Barrier + Sum will slow down compute time so it's definitely a compile time option.

**Comment by manodeep** [comment]
The issue is more that of adding an MPI dependency within the `sage` library. The actual time spent on the barrier, and then the reduction across all nodes should be small compared to the execution time of the rest of the code.
